modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/criteria_builder/ContributorsBuilder.es.js

```diff
@@ -0,0 +1,282 @@
+import Conjunction from './Conjunction.es';
+
+class ContributorsBuilderComp extends React.Component {
+	constructor(props) {
+		super(props);
+		this.classes = getCN(
+			'criteria-builder-root',
+			{
+				'read-only': false
+			}
+		);
+		const {initialContributors, propertyGroups} = props;
+		const contributors = initialContributors && initialContributors.map(
+			c => {
+				const propertyGroup = propertyGroups && propertyGroups.find(t => c.propertyKey === t.propertyKey);
+
+				return {
+					conjunctionId: c.conjunctionId || 'and',
+					conjunctionInputId: c.conjunctionInputId,
+					criteriaMap: c.initialQuery ?
+						translateQueryToCriteria(c.initialQuery) :
+						null,
+					inputId: c.inputId,
+					modelLabel: propertyGroup && propertyGroup.name,
+					properties: propertyGroup && propertyGroup.properties,
+					propertyKey: c.propertyKey,
+					query: c.initialQuery
+				};
+			}
+		);
+		this.onCriteriaEdit = this.onCriteriaEdit.bind(this);
+		this.onCriteriaChange = this.onCriteriaChange.bind(this);
+		this.state = {
+			conjunctionName: 'and',
+			contributors,
+			editing: undefined,
+			newPropertyKey: this.props.propertyGroups.length &&
+				this.props.propertyGroups[0].propertyKey
+		};
+		this._handleRootConjunctionClick = this._handleRootConjunctionClick.bind(this);
+		this._handleSelectorChange = this._handleSelectorChange.bind(this);
+	}
+
+	onCriteriaEdit(id, editing) {
+		this.setState(
+			{
+				editing: editing ? undefined : id
+			}
+		);
+	}
+
+	_handleSelectorChange(e) {
+		const newPropertyKey = e.target.value;
+
+		this.setState(
+			prevState => (
+				{
+					...prevState,
+					newPropertyKey
+				}
+			)
+		);
+	}
+
+	onCriteriaChange(criteriaChange, index) {
+		this.setState(
+			state => {
+				let newState;
+				if (state.editing !== index) {
+					newState = state;
+				}
+				else {
+					newState = {
+						contributors: state.contributors.map(
+							(c, i) => {
+								return (index === i) ? {
+									...c,
+									criteriaMap: criteriaChange,
+									query: buildQueryString([criteriaChange])
+								} : c;
+							}
+						)
+					};
+				}
+				return newState;
+			}
+		);
+	}
+
+	_handleRootConjunctionClick(event) {
+		event.preventDefault();
+
+		this.setState(
+			(prevState, prevProps) => {
+				const prevContributors = prevState.contributors;
+
+				const prevConjunction = prevContributors[0] && prevContributors[0].conjunctionId;
+
+				const {supportedConjunctions} = prevProps;
+				const conjunctionIndex = supportedConjunctions.findIndex(
+					item => item.name === prevConjunction
+				);
+				const conjunctionSelected = (conjunctionIndex === supportedConjunctions.length - 1) ?
+					supportedConjunctions[0].name :
+					supportedConjunctions[conjunctionIndex + 1].name;
+
+				const contributors = prevContributors.map(
+					c => ({
+						...c,
+						conjunctionId: conjunctionSelected
+					})
+				);
+
+				return {
+					...prevState,
```

Comment:
Why are we spreading in the previous state? `setState` will only update the specified state, so there is no need to provide the old state.

For example these two setState functions would result in the same thing.

```js
// state = {one: 1, two: 2}
this.setState(() => ({one: 2}))

// no need to write it this way
this.setState(() => ({one: 2, two: 2}))

// Both result in
// state = {one: 2, two: 2}
```

Reply:
Missed this comment too. Totally right, my bad. I'm too used to reducers.